clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -634,12 +634,12 @@ public boolean matches(ClientRequest request) {
 
 
     @Test
-    public void testExcludeInternalTopicsConfigOption() { 
+    public void testExcludeInternalTopicsConfigOption() {
         subscriptions.subscribe(Pattern.compile(".*"), rebalanceListener);
 
         metadata.update(TestUtils.singletonCluster(TopicConstants.GROUP_METADATA_TOPIC_NAME, 2), time.milliseconds());
 
-        assertFalse(subscriptions.partitionAssignmentNeeded());
+        assertTrue(subscriptions.partitionAssignmentNeeded());
```

Comment:
We handle this case in `ConsumerCoordinator` I think. We basically store a snapshot of the metadata and request a rebalance if anything has changed (including changes to the number of partitions). Now that I'm thinking about it, this is probably why we didn't set `needsPartitionAssignment` to true when the regex changed. We were probably waiting for the metadata refresh to see if we needed to rebalance or not. I'm actually wondering if all that we were really missing was the call to `this.metadata.requestUpdate()` which you added above.